### DISCUSSION TITLE
datetime inspects as UTC string, but json.encode's undesirably as int

### DISCUSSION THREAD

**Discussion by matthewkooshad** [simolus3/drift#2486] [Q&A] [answered]
`DateTimeColumn get insertTime => dateTime().withDefault(Constant(DateTime.now()))();`

when i inspect upon insert, i see:
`DateTime (2023-06-16 23:09:11.178512)`

when i json.encode the object (instance of `class Stuff extends DataClass implements Insertable<Stuff>`), i see:
`"{"id":0,"title":"nothing","insertTime":1686974951178}"`

how can i get the `2023-06-16 23:09:11.178512` value in the json instead? thanks!

**Comment by simolus3** [comment, marked as answer]
Drift actually stores date times as unix timestamps by default, although [this can be changed](https://drift.simonbinder.eu/docs/getting-started/advanced_dart_tables/#datetime-options). The `DateTime (2023-06-16 23:09:11.178512)` value that you're seeing is just the result of `DateTime.toString()` because we're using a `DateTime` object in the row class.

>how can i get the `2023-06-16 23:09:11.178512` value in the json instead? thanks!

Do this before you call the serialization method:

```dart
  driftRuntimeOptions.defaultSerializer = ValueSerializer.defaults(
    serializeDateTimeValuesAsString: true,
  );
```

**Comment by matthewkooshad** [comment]
thanks for the help on this, and it helped me moved forward. 

in utilizing it, i'm facing a situation where i am having a mismatch of values that both say they are utc.

my **API** returns from a remote DB:  `"lastChange": "2023-06-24T18:05:55.1650692-05:00"`
which when this gets mapped to the Dart object, it is understood as: `DateTime (2023-06-24 18:04:05.375878Z)`

my **flutter** app after pulling from the **sqlite** to compare has: `DateTime (2023-06-24 23:04:05.375878Z)`

notice that the hours are a 5 hour difference. i haven't been able to figure this out, to know if there's further tweaking needed.

currently, my system where i am running the flutter app, it is 6pm (reflecting the `"2023-06-24T18:05:55.1650692-05:00"` from the API).

any thoughts on this? thanks!

**Reply by simolus3** [reply]
> notice that the hours are a 5 hour difference

There isn't, the two date times are equal. Dart's `DateTime` doesn't support arbitrary offsets, so you can either get local or UTC date times. As you are five hours behind UTC, the UTC version of your date time appears to be five hours later. You can use `DateTime.toUtc()` or `.toLocal()` to convert between the two.

It's still weird because drift (when not storing date times as text) will return local date times, so I'm not sure why the UTC version has been picked here.

**Reply by matthewkooshad** [reply]
Thanks, I'll give it a try. I just wanted to give a quick response, that the datetime stored is coming from my api and then stored in sqlite. The value was not generated in dart.

**Reply by matthewkooshad** [reply]
what i found worked for me:
1. in flutter app's build.yaml
```
targets:
  $default:
    builders:
      drift_dev:
        options:
          store_date_time_values_as_text: true
```
2. in dart, when creating the db instance when flutter app loads:
```
driftRuntimeOptions.defaultSerializer = const ValueSerializer.defaults(
    serializeDateTimeValuesAsString: true,
  );
```
3. in dart, adjusting my drift sqlite table schema to use Text instead:
`TextColumn get lastChange => text()();`
4. in dart, when i store the present datetime, i use:
`DateTime.now().toIso8601String()`
5. in my C# API, when i store the present datetime, i use:
`DateTime.Now.ToUniversalTime();`
6. in my C# API, i have this in my db context, which will return values like this `"lastChange": "2023-06-29T16:41:18.4446871Z",` instead:
```
protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
    {
      configurationBuilder
          .Properties<DateTime>()
          .HaveConversion(typeof(UtcValueConverter));
    }

    class UtcValueConverter : ValueConverter<DateTime, DateTime>
    {
      public UtcValueConverter()
          : base(v => v, v => DateTime.SpecifyKind(v, DateTimeKind.Utc))
      {
      }
    }
```
7. in dart, when i need to compare with datetime values that are pulled from my API:
`DateTime.parse`
